### PR TITLE
[JIT] during freezing, cast optional bias to half if weight is half

### LINKED ISSUE #76578: [JIT][Autocast] Batchnorm folding pass during freezing doesn't preserve types
### 🐛 Describe the bug

The optimize_numerics pass in the freezing pass interferes with the types used by the autocasting freezing pass.

Autocast() during freezing will cast weights to half precision; but then FoldFrozenConvBatchnorm turns them into full precision, causing errors when the model is run. 

```python
import torch
from typing import Final
net = torch.nn.Sequential(torch.nn.Conv2d(3, 64, (3, 3), 1, 1, bias=False),
                                      torch.nn.BatchNorm2d(64),
                                      torch.nn.PReLU(64)).cuda().eval()
img = torch.randn(3, 3, 112,112).cuda()
with torch.cuda.amp.autocast(True):
    net = torch.jit.trace(net, img, check_trace=False)
    net = torch.jit.freeze(net)
print(net.graph)
feature = net(img)  # forward pass
```

Workaround for now - you can disable optimize_numerics during freezing with `torch.jit.freeze(model, optimize_numerics=False)`

thanks to @xsacha who found the issue & provided the repro

### Versions

master

### PR DESCRIPTION
**Pull request by davidberard98**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #77295

On GPU, conv errors if not all its inputs have the same dtype.

In the case of autocasting during freezing, what we see is:
1) inputs to conv are casted to half
2) inputs to batchnorm are not casted, so many are still floats
3) we try to fold conv + batchnorm, by finding different weight and bias such that conv(input, new_weight, new_bias) is equivalent to the original conv -> batchnorm.

If conv previously had an optional bias, then during freezing we will temporarily create a zero-valued bias as a placeholder for conv_bias. We want to construct it to have the same dtype as the weight input to conv, to avoid errors on GPU.

Fixes #76578

### CODE REVIEW

**Comment by davidberard98** [comment]
cc @ngimel this is the test where we're seeing the cuda caching allocator failures:
```
$ python test/test_jit.py -k test_conv_bn_folding_autocast
... caching allocator leaks 2560 ...
$ PYTORCH_NO_CUDA_MEMORY_CACHING=1 python test/test_jit.py -k test_conv_bn_folding_autocast
... succeeds ...
```

File: test/jit/test_freezing.py [hunk 1]
```diff
@@ -1678,6 +1678,30 @@ def make_prediction(self, x):
...
+                self.conv = torch.nn.Conv2d(in_channels, out_channels, bias=False, dtype=torch.half, **kwargs)
+                self.bn = torch.nn.BatchNorm2d(out_channels, eps=0.001, dtype=torch.float)
+
+            def forward(self, x):
+                return self.bn(self.conv(x))
+
+        mod_eager = ConvBN(3, 32, kernel_size=3, stride=2).cuda().eval()
+        scripted_mod = torch.jit.script(mod_eager)
+        scripted_mod = torch.jit.freeze(scripted_mod)
+        FileCheck().check("conv").check_not("aten::batch_norm").run(scripted_mod.graph)
```
Comment (eellison) [inline test/jit/test_freezing.py:1698]: Could you add a check that the `conv` weight here is `float` ?

File: torch/csrc/jit/passes/frozen_conv_folding.cpp [hunk 2]
```diff
@@ -79,7 +79,20 @@ void FoldFrozenConvBatchnorm(Block* b) {
...
-        conv_b = at::zeros_like(bn_rm);
+        // If this is on GPU and bias is none and weight was half/bfloat, but
+        // bn_rm was float, then probably this was a case where autocasting
+        // casted inputs to conv. And since CUDA conv implementation requires
+        // all the inputs to have the same scalar dtype, we need to make this
+        // placeholder have the same type as conv_w.
+        at::ScalarType bias_dtype = bn_rm.scalar_type();
+        at::ScalarType weight_dtype = conv_w.scalar_type();
+        at::DeviceType weight_device = conv_w.device().type();
+        if (weight_device == at::kCUDA &&
```
Comment (eellison) [inline torch/csrc/jit/passes/frozen_conv_folding.cpp:90]: is this any reason to not just always put `conv_b` as the same dtype as `weight_dtype` ? for cpu as well ?

**Review by eellison — APPROVED** [review]
LGTM! Just a couple of small questions

File: test/jit/test_freezing.py (hunk 1 above)
Comment (eellison) [reply test/jit/test_freezing.py:1698]: sorry, `conv` weight is `half`

File: torch/csrc/jit/passes/frozen_conv_folding.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (davidberard98) [reply torch/csrc/jit/passes/frozen_conv_folding.cpp:90]: on CPU this won't error out, so maybe there's some use case where users want bias to be a float.

^ but if you think that's not really an issue, it's probably fine to always cast

**Review by eellison — APPROVED** [review]
